contracts/governance/FeeSharingCollector/FeeSharingCollector.sol

```diff
@@ -373,76 +457,198 @@ contract FeeSharingCollector is
+        nonReentrant
+    {
+        // @dev e.g. _fromCheckpoint == 10
+        // after _withdraw() user's processedCheckpoints should be 10 =>
+        // set processed checkpoints = 9, next maping index = 9 (10th checkpoint)
+        uint256 prevFromCheckpoint = _fromCheckpoint.sub(1);
+        if (
+            prevFromCheckpoint >
+            processedCheckpoints[msg.sender][RBTC_DUMMY_ADDRESS_FOR_CHECKPOINT]
+        ) {
+            processedCheckpoints[msg.sender][
+                RBTC_DUMMY_ADDRESS_FOR_CHECKPOINT
+            ] = prevFromCheckpoint;
+        }
+        _withdrawRBTC(_maxCheckpoints, _receiver);
+    }
+
+    /**
+     * @dev Returns first user's checkpoint with weighted stake > 0
+     *
+     * @param _user The address of the user or contract.
+     * @param _token RBTC dummy to fit into existing data structure or SOV. Former address of the pool token.
+     * @param _startFrom Checkpoint number to start from. If _startFrom < processedUserCheckpoints then starts from processedUserCheckpoints.
+     * @param _maxCheckpoints Max checkpoints to process in a row to avoid timeout error
+     * @return [checkpointNum: checkpoint number where user's weighted stake > 0, hasSkippedCheckpoints, hasFees]
+     */
+    function getNextPositiveUserCheckpoint(
+        address _user,
+        address _token,
+        uint256 _startFrom,
+        uint256 _maxCheckpoints
+    )
+        external
+        view
+        returns (
+            uint256 checkpointNum,
+            bool hasSkippedCheckpoints,
+            bool hasFees
+        )
+    {
+        if (staking.isVestingContract(_user)) {
+            return (0, false, false);
+        }
+        require(_maxCheckpoints > 0, "_maxCheckpoints must be > 0");
+
+        uint256 totalCheckpoints = totalTokenCheckpoints[_token];
+        uint256 processedUserCheckpoints = processedCheckpoints[_user][_token];
+
+        if (processedUserCheckpoints >= totalCheckpoints || totalCheckpoints == 0) {
+            return (totalCheckpoints, false, false);
+        }
+
+        uint256 startFrom =
+            _startFrom > processedUserCheckpoints ? _startFrom : processedUserCheckpoints;
+
+        uint256 end = startFrom.add(_maxCheckpoints);
+        if (end >= totalCheckpoints) {
+            end = totalCheckpoints;
+        }
+
+        // @note here processedUserCheckpoints is a number of processed checkpoints and
+        // also an index for the next checkpoint because an array index starts wtih 0
+        for (uint256 i = startFrom; i < end; i++) {
+            Checkpoint storage tokenCheckpoint = tokenCheckpoints[_token][i];
+            uint96 weightedStake =
+                staking.getPriorWeightedStake(
+                    _user,
+                    tokenCheckpoint.blockNumber - 1,
+                    tokenCheckpoint.timestamp
+                );
+            if (weightedStake > 0) {
+                // i is the index and we need to return checkpoint num which is i + 1
```

Comment:
would it be simpler to just return the index? then, we wouldn't need to subtract 1 from the passed param on the withdraw functions, no?

Reply:
by design most of the functions return checkpoints number rather then index. it looks weird at first but it makes sense because we mostly need to work the next checkpoint, which is the processedCheckpoints[user][token] - index and number of processed tokens at the same time. therefore we either just use the returned values for most of the cases as index for the next checkpoint or add 1 every time. both are fine, i'm just sticking with the existing solution.